spacy/tokens/doc.pyx

```diff
@@ -1451,6 +1451,140 @@ cdef class Doc:
                 remove_label_if_necessary(attributes[i])
                 retokenizer.merge(span, attributes[i])
 
+    def from_json(self, doc_json):
+        """Convert a JSON document generated by Doc.to_json() to a Doc.
+
+        vocab (Vocab): A storage container for lexical types.
+        RETURNS (Doc): A doc instance corresponding to the specified JSON representation.
+        """
+
+        ### Token-level properties ###
+
+        words = []
+        annotation_types = ("pos", "head", "dep", "lemma", "tag", "morph")
+        token_annotations = {
+            # For each annotation type: store (1) annotation type required to include it, (2) list of annotations for
+            # this type.
+            annotation_type: {"req": annotation_type if annotation_type != "head" else "dep", "values": None}
+            for annotation_type in annotation_types
+        }
+
+        # Gather token-level properties.
+        for token in doc_json["tokens"]:
+            words.append(doc_json["text"][token["start"]:token["end"]])
+            for annotation_type in token_annotations:
+                # Assuming all tokens have exactly the same set of attributes.
+                if token_annotations[annotation_type]["req"] in token:
+                    if token_annotations[annotation_type]["values"] is None:
+                        token_annotations[annotation_type]["values"] = []
+                    token_annotations[annotation_type]["values"].append(token[annotation_type])
+
+        # Initialize doc instance.
+        start = 0
+        cdef const LexemeC* lex
+        cdef str orth_
+        cdef bint has_space
+        reconstructed_words, spaces = get_words_and_spaces(words, doc_json["text"])
+        assert words == reconstructed_words
+
+        for i, token in enumerate(doc_json["tokens"]):
+            has_space = spaces[i]
+            orth_ = doc_json["text"][token["start"]:token["end"]]
+            lex = self.vocab.get(self.mem, orth_)
+            self.push_back(lex, has_space)
+
+        # Set remaining token-level attributes via Doc.from_array().
+        if token_annotations["head"]["values"] is not None:
+            token_annotations["head"]["values"] = [
+                head - i if head is not None else 0 for i, head in enumerate(token_annotations["head"]["values"])
```

Comment:
I don't think that `head` or `dep` can be `None`? You end up with `"dep":""` or `"head":<id>` for what is effectively missing annotation, so you have rely to `"dep"` to see whether dependency annotation is set.

Reply:
`to_json()` generates the token annotation fields as follows:
```python
if include_annotation["DEP"]:
    token_data["dep"] = token.dep_
    token_data["head"] = token.head.i
```
In my understanding: if the coument to be converted to JSON doesn't have the DEP annotation, the token data won't include either "dep" nor "head" entries. Consequently at this point in `from_json()` `token_annotations["head"]["values"]` respectively `token_annotations["dep"]["values"]` will be `None`, since they couldn't been collected from the document JSON.

Reply:
I think that `token_annotations["dep"]["values"]` is `None`, but the individual values (if they are set) have no way of being `None`. `"dep"` will always be a string and `"head"` will always be an int.

Reply:
I see, so I a proper check would be like
```python
head - i if token_annotations["dep"]["values"] is != "" else 0 for i, head in enumerate(token_annotations["head"]["values"]) 
```
Also - this is handled the same way in `__init__()`. Should it be changed accordingly there too?

Reply:
Got it. Is the check for an empty string in individual `"dep"` values as shown in the last comment what you meant?

Reply:
`token_annotations["dep"]["values"]` is checking the list value, so I think you meant `token_annotations["dep"]["values"][i] != ""` (not `is !=`)? (I considered this to be part of "how to handle partial dependency annotation".)